x/incentive/README.md

```diff
@@ -0,0 +1,153 @@
+   StartTime: 1679659746, // Mar 24 2023
+   Duration: 864000, // 10 days
+   UToken: "u/uumee",
+   TotalRewards: sdk.Coin{"uumee",1000_000000},
+}
+```
+
+Reward distribution math is
+- **Constant Rate:** Regardless of how much `u/uumee` is bonded to the incentive module at any given block, the program distributes the fraction `blockTime / duration` of its total rewards across users every block it is active (with some corrective rounding).
+- **Weighted by Amount:** A user with `200u/uumee` bonded received twice as many rewards on a given block as a user with `100u/uumee` bonded.
+- **Weighted by Tier:** A uToken bonded to the short or middle unbonding tier receives fewer rewards than one bonded to the long tier. This exact ratio is determined by the `TierWeightMiddle` and `TierWeightShort` parameters.
+
+When multiple incentive programs are active simultaneously, they compute their rewards independently.
+
+### Claiming Rewards
+
+A user can claim rewards for all of their bonded uTokens at once using `MsgClaim`. When a user claims rewards, an appropriate amount of liquid tokens are sent from the `x/incentive` module account to their wallet.
+
+There are also times where rewards must be claimed automatically to maintain rewards-tracking math. These times are:
+- On `MsgBond`
+- On `MsgBeginUnbonding`
+- When a `leverage.MsgLiquidate` forcefully reduces the user's bonded collateral
+
+Any of the actions above cause the same tokens to be transferred to the user as would have been generated by a `MsgClaim` at the same moment, or a subset of them if only claiming rewards asociated with a single bonded uToken denom or tier.
+
+By automatically claiming rewards whenever a user's bonded amount changes, the module guarantees a mathematically useful invariant:
+
+> Between any two consecutive reward claims by an account associated with a specific bonded uToken denom and tier, the amount of bonded `uTokens` of the given denom to that tier remained constant.
+
+### Reward Accumulators
+
+The incentive module must calculate rewards each block without iterating over accounts and past incentive programs. It does so by storing a number of `RewardAccumulator`
+
+```go
+ra := RewardAccumulator{
+   denom: "u/uumee",
+   exponent: 6,
+   tier: 2, // BondTierMiddle
+   rewards: "0.00023uumee, 0.00014ibc/1234",
+}
+```
+
+The example reward accumulator above can be interpreted as:
+
+> If `10^6` `u/uumee` was bonded to tier `2` at genesis and had remained bonded since, it would have accumulated `0.00023uumee` and `0.00014ibc/1234` in rewards.
+
+The incentive module must store 3 `RewardAccumulator` for each bonded uToken denom, since each reward accumulator specifies only a single tier.
+
+During `EndBlock` when rewards are being distributed by incentive programs, the module divides the amount of tokens to distribute by the current `TotalBonded` which it stores for each uToken denom and tier, to increment the `rewards` field in each `RewardAccumulator`.
```

Comment:
It would be required every time any user does a `MsgBond`, `MsgBeginUnbonding`, `MsgClaim`, or `leverage.MsgLiquidate`, or when an incentive program began or ended or module params changed. It would also have to be simulated (but not updated) any time someone queried pending rewards.

(in other words, any time `TotalBonded` changes plus edge cases)

That might save a little on iteration, but since the current approach only iterates over `OngoingIncentivePrograms`, not denoms or users, it seems already to be fast enough to avoid the complexity above.

Reply:
Which is much better than doing it every block.

Reply:
Not necessarily better - it replaces

> once per block

with

> once every time one of 6 cases occurs, which may be a 3 one of three transaction types, or a leverage hook, or a governance action, or an endblock [on program end], _potentially multiple times per block_

The complexity isn't worth it and it actually scales worse during transaction spam

Reply:
If everyone will do it once per block then we get to the same complexity. Most of the users won't do it every block, so we have performance gain.

Reply:
Isn't this as simple as adding the update function to every operation which changes the program participation?